src/TheSpace/HarbergerMarket.sol

```diff
@@ -69,120 +84,177 @@ abstract contract HarbergerMarket is Multicall {
-        property = new Property(propertyName_, propertySymbol_, address(this), totalSupply_);
+        property = new Property(propertyName, propertySymbol, address(this), totalSupply);
 
         // initialize currency contract
-        currency = ERC20(currencyAddress_);
+        currency = ERC20(currencyAddress);
 
-        // TODO: tax rate setter
-        totalSupply = totalSupply_;
-        taxRate = taxRate_;
+        // default config
+        taxConfig[ConfigOptions.taxRate] = 10;
+        taxConfig[ConfigOptions.treasuryShare] = 500;
     }
 
     /**
      * @dev Set the current price of an Harberger property with token id.
      *
      * Emits a {Price} event.
      */
-    function setPrice(uint256 tokenId_, uint256 price_) external {
-        require(property.ownerOf(tokenId_) == msg.sender, "Sender does not own property");
+    function setTaxConfig(ConfigOptions option, uint256 value) external onlyOwner {
+        taxConfig[option] = value;
 
-        _setPrice(tokenId_, price_);
+        emit Config(option, value);
+    }
+
+    // TODO: withraw community treasury
+
+    /**
+     * @dev Set the current price of an Harberger property with token id.
+     *
+     * Emits a {Price} event.
+     */
+    function setPrice(uint256 tokenId, uint256 price) external {
+        if (!property.isApprovedOrOwner(msg.sender, tokenId)) revert Unauthorized();
+        if (price == this.getPrice(tokenId)) return;
+
+        bool success = this.collectTax(tokenId);
+        if (success) _setPrice(tokenId, price);
     }
 
     /**
      * @dev Returns the current price of an Harberger property with token id.
      */
-    function getPrice(uint256 tokenId_) external view returns (uint256 price) {
-        return taxRecord[tokenId_].price;
+    function getPrice(uint256 tokenId) external view returns (uint256 price) {
+        return taxRecord[tokenId].price;
+    }
+
+    /**
+     * @dev Returns the current owner of an Harberger property with token id.
+     */
+    function getOwner(uint256 tokenId) external view returns (address owner) {
+        return property.ownerOf(tokenId);
+    }
+
+    /**
+     * @dev calculate tax for a token
+     */
+    function getTax(uint256 tokenId) external view returns (uint256) {
+        // calculate tax
+        // 1000 for every 1000 blocks, 10000 for conversion from bps
+        return
+            (this.getPrice(tokenId) *
+                taxConfig[ConfigOptions.taxRate] *
+                (block.number - taxRecord[tokenId].lastTaxCollection)) / (1000 * 10000);
     }
 
     /**
      * @dev Purchase property with bid higher than current price. Clear tax for owner before transfer.
      * TODO: check security implications
      */
-    function bid(uint256 tokenId_, uint256 price_) external {
-        // TODO: mint token if not already exists
+    function bid(uint256 tokenId, uint256 price) external {
+        if (property.exists(tokenId)) {
```

Comment:
looks we did a lot of repetitive checks on token ownership: `exists`, `ownerof`, `getPrice`, `collectTax`